modules/frontend/querysharding_test.go

```diff
@@ -52,129 +60,217 @@ func TestCreateBlockShards(t *testing.T) {
 	}
 }
 
-func TestMergeResponses(t *testing.T) {
-	t1 := test.MakeTrace(10, []byte{0x01, 0x02})
-	t2 := test.MakeTrace(10, []byte{0x01, 0x03})
+func TestBuildShardedRequests(t *testing.T) {
+	queryShards := 2
 
-	b1, err := proto.Marshal(t1)
-	assert.NoError(t, err)
-	b2, err := proto.Marshal(t2)
-	assert.NoError(t, err)
+	sharder := &shardQuery{
+		queryShards:     queryShards,
+		blockBoundaries: createBlockBoundaries(queryShards - 1),
+	}
 
-	combinedTrace, _, err := model.CombineTraceBytes(b1, b2, model.TracePBEncoding, model.TracePBEncoding)
-	assert.NoError(t, err)
+	ctx := user.InjectOrgID(context.Background(), "blerg")
+	req := httptest.NewRequest("GET", "/", nil).WithContext(ctx)
+
+	shardedReqs, err := sharder.buildShardedRequests(req)
+	require.NoError(t, err)
+	require.Len(t, shardedReqs, queryShards)
+
+	require.Equal(t, "/querier/?mode=ingesters", shardedReqs[0].RequestURI)
+	require.Equal(t, "/querier/?blockEnd=ffffffffffffffffffffffffffffffff&blockStart=00000000000000000000000000000000&mode=blocks", shardedReqs[1].RequestURI)
+}
+
+func TestShardingWareDoRequest(t *testing.T) {
+	// create and split a trace
+	trace := test.MakeTrace(10, []byte{0x01, 0x02})
+	trace1 := &tempopb.Trace{}
+	trace2 := &tempopb.Trace{}
+
+	for _, b := range trace.Batches {
+		if rand.Int()%2 == 0 {
+			trace1.Batches = append(trace1.Batches, b)
+		} else {
+			trace2.Batches = append(trace2.Batches, b)
+		}
+	}
 
 	tests := []struct {
-		name            string
-		requestResponse []RequestResponse
-		expected        *http.Response
+		name           string
+		status1        int
+		status2        int
+		trace1         *tempopb.Trace
+		trace2         *tempopb.Trace
+		err1           error
+		err2           error
+		expectedStatus int
+		expectedTrace  *tempopb.Trace
+		expectedError  error
 	}{
 		{
-			name: "combine status ok responses",
-			requestResponse: []RequestResponse{
-				{
-					Response: &http.Response{
-						StatusCode: http.StatusOK,
-						Body:       io.NopCloser(bytes.NewReader(b1)),
-					},
-				},
-				{
-					Response: &http.Response{
-						StatusCode: http.StatusOK,
-						Body:       io.NopCloser(bytes.NewReader(b2)),
-					},
-				},
-				{
-					Response: &http.Response{
-						StatusCode: http.StatusNotFound,
-						Body:       io.NopCloser(bytes.NewReader([]byte("foo"))),
-					},
-				},
-			},
-			expected: &http.Response{
-				StatusCode:    http.StatusOK,
-				Body:          io.NopCloser(bytes.NewReader(combinedTrace)),
-				ContentLength: int64(len(combinedTrace)),
-			},
+			name:           "empty returns",
```

Comment:
Nice refactoring! I think this needs one more case: Any internal error code other than 404 should be translated into 500

Reply:
Good catch. Forgot about that behavior. Fixed.